### ISSUE TITLE
Channels: follow-up work

### ISSUE THREAD

**Issue by iravid** [zio/zio#4886]
After #4885 is merged, we have the following open issues that need to be resolved before it can replace the implementation in `zio.stream`:
* Channels
	* Decide what to do with the Bridge node
	* Add a suspend operator so we can implement combinators without `Ref`
	* Implement n-ary merge
* Sinks
	* Improve inference problems with InErr, In
	* Combinators
		* collectAllWhileWith
		* foldM - preserve leftovers on failure
		* raceBoth
		* timed
		* repeat
		* zipWithPar
		* untilOutputM
	* Constructors
		* managed
	* Port from ZTransducer:
		* filterInput/M
		* branchAfter
		* collectAllN
		* collectAllToMapN
		* collectAllToSetN
		* dropWhile/M
		* iso_8859_1Decode
		* splitLines
		* splitOn
		* splitOnChunk
		* utfDecode
		* utf8Decode
		* utf16Decode
		* utf16BEDecode
		* utf16LEDecode
		* utf32Decode
		* utf32BEDecode
		* utf32LEDecode
		* usASCIIDecode
		* deflate
		* inflate
		* gzip
		* gunzip
* ZStream
	* Combinators
		* Rewrite broadcast* distributed* and groupBy on top of ZHub (where possible)
		* bufferDropping
		* bufferSliding
		* combine without pull
		* combineChunks without pull
		* drop swallows errors
		* dropWhile swallows errors
		* decide what to do with ensuringFirst
		* use sinks in all the run* methods and do not implement functionality on ZStream
		* Implement flatMapParSwitch somehow (maybe push this functionality into Channel’s n-ary merge)
		* haltWhen
		* grouped
		* groupedWithin
		* interleaveWith
		* intersperse
		* mapAccumM (use loopOnPartial)
		* mapError (through channel)
		* mapErrorCause (through channel)
		* mapMPar (probably through channel)
		* mergeWith (through channel)
		* peel without Pull
		* refineOrDie
		* refineOrDieWith
		* repeatElementsWith
		* repeatWith
		* runLast
		* runCount
		* runSum
		* scanReduceM (loopOnPartial)
		* scheduleWith
		* takeRight
		* throttleEnforceM (package functionality into a sink)
		* throttleShapeM (package functionality into a sink)
		* debounce (sink)
		* toInputStream
		* toIterator
		* toReader
		* zipWith and zipAllWithExec without pull, with Adam’s specialized zip RingBuffer
		* zipWithLatest
		* zipWithNext
	* Constructors
		* paginate without going through paginateM
		* paginateChunk without going through paginateChunkM
		* repeat without going through repeatEffect
		* unfold without going through unfoldM
		* unfoldChunk without going through unfoldChunkM
		* effectAsync*

**Comment by mijicd** [comment]
Let's start shy then, I'm taking paginate* and unfold* :smile:

**Comment by paulpdaniels** [comment]
I'll take these:

- mapAccumM (use loopOnPartial)
- mapError (through channel)
- mapErrorCause (through channel)

**Comment by paulpdaniels** [comment]
I also did `scheduleWith` and `scanReduceM` since they were pretty straight forward

**Comment by mijicd** [comment]
I'm taking `grouped` and `groupedWithin`.

**Comment by iravid** [comment]
Taking utf8Decode

**Comment by paulpdaniels** [comment]
@iravid I just started with a `branchAfter` implementation is that needed for the `ut*` variants?

**Comment by iravid** [comment]
@paulpdaniels Hey it turns out that `branchAfter` is easily implemented as `ZSink.take(n).flatMap { data =>`. So probably not.

**Comment by mijicd** [comment]
I'm taking `ZSink.collectAllN` port as it's needed in grouped*.

**Comment by paulpdaniels** [comment]
Just to update here as well:

- ZStream#throttleShapedM
- ZStream#throttleEnforcedM
- ZSink#timed (+ZSink#summarized)
- use sinks in all the run* methods and do not implement functionality on ZStream
- runLast
- runCount
- runSum

**Comment by paulpdaniels** [comment]
Took `repeatElementsWith` as well

**Comment by paulpdaniels** [comment]
Starting on `debounce`

**Comment by mijicd** [comment]
I'll take `ZSink.collectAllToMapN` and `ZSink.collectAllToSetN`.

**Comment by luis3m** [comment]
I'll take bufferSliding and bufferDropping

**Comment by mijicd** [comment]
Taking:
- `ZSink.splitLines`
- `ZSink.splitOn`
- `ZSink.splitOnChunk`

`ZStream.grouped` and `ZStream.groupedWithin` are ready, I'll open the PR as soon as #4923 is merged.

**Comment by paulpdaniels** [comment]
Taking `drop` and `dropWhile`

**Comment by paulpdaniels** [comment]
Taking `peel` and `repeatWith`

**Comment by adamgfraser** [comment]
I will take `combine`, `combineChunks`, and `interleaveWith`.

**Comment by paulpdaniels** [comment]
Taking `repeat`

**Comment by luis3m** [comment]
Taking `zipWithNext`

**Comment by luis3m** [comment]
By the way, for whomever takes (I will probably after `zipWithNext`) `mapMPar` here's a known issue: #4514 (#4731 fix for ZIO 1.x)

**Comment by sviezypan** [comment]
I will take 
```
ZSink#collectAllWhileWith
ZSink#untilOutputM
```

**Comment by iravid** [comment]
I'll be back reviewing tomorrow, sorry for the delay!

**Comment by luis3m** [comment]
Taking `zipWithLatest`

**Comment by jupposessho** [comment]
Taking filterInput/filterInputM

**Comment by agarella** [comment]
Taking `toIterator`

**Comment by zagyi** [comment]
Taking `toInputStream`

**Comment by luis3m** [comment]
Taking `mergeWith`

**Comment by luis3m** [comment]
Taking `mapMPar`

**Comment by paulpdaniels** [comment]
Taking `ZSink#dropWhile/M`

**Comment by paulpdaniels** [comment]
Taking `effectAsync*`

**Comment by jkobejs** [comment]
I took 
- deflate
- gzip

**Comment by paulpdaniels** [comment]
Can we get this updated? What is remaining?

**Comment by mijicd** [comment]
I owe splitOn and splitOnChunk, @jkobejs took splitLines. I could try to update the list for others, but it's better if Itamar tackles that.

**Comment by iravid** [comment]
@paulpdaniels @mijicd I updated the list on Friday. The only thing I missed is the open PR for the remaining ZStream constructors so I'll check those off as well.

I think we basically have only the ZSink stuff left.

**Comment by kmatasflp** [comment]
To familiarize myself with the code base can I take `gunzip` or @iravid there is a better one you recommend me to get started with?

**Comment by jupposessho** [comment]
Taking collectAllWhileWith

**Comment by runtologist** [comment]
Taking `*decode`

**Comment by jupposessho** [comment]
Taking collectAllWhileWith

**Comment by kmatasflp** [comment]
Taking `filterInput/M`

**Comment by jupposessho** [comment]
Taking untilOutputZIO

**Comment by kmatasflp** [comment]
@jupposessho is `untilOutputZIO` you are taking similar to https://github.com/zio/zio/pull/5063 ?

**Comment by jupposessho** [comment]
> @jupposessho is `untilOutputZIO` you are taking similar to #5063 ?

yes, it's the new name of the deprecated method made me confused :)

**Comment by paulpdaniels** [comment]
Taking `zipWithPar`

**Comment by regiskuckaertz** [comment]
Hello 👋 I'd love to help here. Is the list in the OP accurate? I feel like other tickets are tackling some of them, e.g. the charset ones and maybe more.

**Comment by vigoo** [comment]
I think this list was not updated since a while :) Everything that's missing from the old functionality was created as a hackathon card I think

**Comment by mijicd** [comment]
Let's close it then :).
